File: 백준/Gold/2573. 빙산/빙산.py
n, m = map(int, input().split())
board = [list(map(int, input().split())) for _ in range(n)]

from collections import deque, defaultdict

def solution(n, m, board):
    
    dr = [-1, 1, 0, 0]
    dc = [0, 0, -1, 1]
    
    def in_range(r, c):
        return 0 <= r < n and 0 <= c and m
    
    def find_ice(): # 빙하 판단
        
        in_queue = [[False for _ in range(m)] for _ in range(n)]
        
        def bfs(r, c):
            dq = deque([])
            
            dq.append((r, c))
            in_queue[r][c] = True
            ice = [(r, c)]
            
            while dq:
                cr, cc = dq.popleft()
                for k in range(4):
                    nr = cr + dr[k]
                    nc = cc + dc[k]
                    if in_range(nr, nc) and\
                        not in_queue[nr][nc] and\
                            board[nr][nc] != 0:
                                dq.append((nr, nc))
                                in_queue[nr][nc] = True
                                ice.append((nr, nc))
            return ice
        
        ices = defaultdict(list)
        cnt = 1
        for i in range(n):
            for j in range(m):
                if in_queue[i][j] or board[i][j] == 0:
                    continue
                ice = bfs(i, j)
                ices[cnt] = ice
                cnt += 1
    
        return ices

    def melt(ices): 
        # ices = find_ice().values()
        check_water = []
        for ice in ices:
            for r, c in ice:
                cnt = 0
                for k in range(4):
                    nr = r + dr[k]
                    nc = c + dc[k]
                    if in_range(nr, nc) and\
                        board[nr][nc] == 0:
                            cnt += 1
                check_water.append(cnt)
        
        # print(ices)
        # print(check_water)
        
        for ice in ices:
            for (r, c), cnt in zip(ice, check_water):
                board[r][c] = max(board[r][c] - cnt, 0)

    
    time = 0
    while True:
        ices = find_ice()   # {
                            #    1 : [(r1, c1)...]
                            #  , 2 : [(r2, c2)..]]
                            # }
                            
        if len(ices.keys()) == 0:
            return 0
        
        if len(ices.keys()) >= 2:
            return time
        
        time += 1
        melt(ices.values())
    

print(solution(n, m, board))

File: 백준/Gold/2573. 빙산/빙산.java
import java.io.*;
import java.util.*;

class Solution {
    static int N, M;
    static int[][] board;
    static int[] dr = {-1, 1, 0, 0};
    static int[] dc = {0, 0, -1, 1};

    public static int solution(int N, int M, int[][] board) {
        Solution.N = N;
        Solution.M = M;
        Solution.board = board;

        int t = 0;
        while (true) {
            IceResult result = findIces();
            List<int[]> ices = result.ices;
            int cnt = result.cnt;

            if (cnt == 0) {
                return 0;
            }
            if (cnt >= 2) {
                return t;
            }

            event(ices);
            t++;
        }
    }

    static boolean inRange(int r, int c) {
        return 0 <= r && r < N && 0 <= c && c < M;
    }

    static List<int[]> bfs(int r, int c, boolean[][] inQueue) {
        List<int[]> ices = new ArrayList<>();
        Queue<int[]> dq = new LinkedList<>();

        dq.offer(new int[]{r, c});
        ices.add(new int[]{r, c});
        inQueue[r][c] = true;

        while (!dq.isEmpty()) {
            int[] cur = dq.poll();
            int cr = cur[0];
            int cc = cur[1];

            for (int k = 0; k < 4; k++) {
                int nr = cr + dr[k];
                int nc = cc + dc[k];

                if (inRange(nr, nc) && !inQueue[nr][nc] && board[nr][nc] != 0) {
                    dq.offer(new int[]{nr, nc});
                    ices.add(new int[]{nr, nc});
                    inQueue[nr][nc] = true;
                }
            }
        }

        return ices;
    }

    static class IceResult {
        List<int[]> ices;
        int cnt;

        IceResult(List<int[]> ices, int cnt) {
            this.ices = ices;
            this.cnt = cnt;
        }
    }

    static IceResult findIces() {
        List<int[]> ices = new ArrayList<>();
        boolean[][] inQueue = new boolean[N][M];

        int cnt = 0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (board[i][j] == 0 || inQueue[i][j]) {
                    continue;
                }
                List<int[]> ice = bfs(i, j, inQueue);
                cnt++;
                ices.addAll(ice);
            }
        }

        return new IceResult(ices, cnt);
    }

    static void event(List<int[]> ices) {
        int[][] eventBoard = new int[N][M];

        // 각 얼음 위치에서 인접한 물의 개수만큼 감소값 누적
        for (int[] pos : ices) {
            int r = pos[0];
            int c = pos[1];

            for (int k = 0; k < 4; k++) {
                int nr = r + dr[k];
                int nc = c + dc[k];

                if (inRange(nr, nc) && board[nr][nc] == 0) {
                    eventBoard[r][c] -= 1;  // 수정: 얼음 위치에 감소값 누적
                }
            }
        }

        // 실제 보드에 적용
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (board[i][j] > 0) {
                    board[i][j] += eventBoard[i][j];
                    board[i][j] = Math.max(0, board[i][j]);
                }
            }
        }
    }
}

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        int[][] board = new int[N][M];
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        System.out.println(Solution.solution(N, M, board));
        br.close();
    }
}